Max video file size

chaitanyaraghav opened the pull request:
Hello, 

I had a need to set a max file size limit for my implementation of the camera and added the API to set a limit. I added a new API method `setMaxFileSize(long maxFileSizeInBytes)` to set a custom maximum file size for a video recording. The camera ignores the max file size if its not set or is invalid. 

I added another boolean parameter to the video callback method indicating if the video ended due to a file size limit. I also included a helper getter method to get the status of the current camera video recording. It felt easier as I could rely on `cameraView.isRecordingVideo()` instead of having my own boolean tracking in my activity.

Thanks!

natario1 commented on cameraview/src/main/java/com/otaliastudios/cameraview/Camera1.java:
```diff
@@ -676,6 +678,26 @@ private void initMediaRecorder() {
+                    switch (i){
+                        case MediaRecorder.MEDIA_RECORDER_INFO_MAX_FILESIZE_REACHED:{
+                            mHasReachedMaxSize = true;
+                            endVideo();
```
I think `endVideoImmediately` would work better.

natario1 commented on cameraview/src/main/java/com/otaliastudios/cameraview/Camera1.java:
```diff
@@ -833,4 +855,18 @@ private static Rect computeMeteringArea(double centerX, double centerY, double s
+    // Additional helper info
+
+    @Override
+    boolean isRecordingVideo() {
```
can you move this to `CameraController` and make it `final`?

natario1 commented on cameraview/src/main/java/com/otaliastudios/cameraview/CameraListener.java:
```diff
@@ -73,9 +73,11 @@ public void onPictureTaken(byte[] jpeg) {
      */
     @UiThread
-    public void onVideoTaken(File video) {
+    public void onVideoTaken(File video, boolean hasReachedMaxFileSize) {
```
As said, I think this is unacceptable at the moment. We can't force everyone to rewrite their code. But in the future I will change the signature with a `VideoResult` so we can add everything.

natario1 commented on cameraview/src/main/java/com/otaliastudios/cameraview/CameraView.java:
```diff
@@ -1346,7 +1364,7 @@ public boolean getPlaySounds() {
         void processImage(byte[] jpeg, boolean consistentWithView, boolean flipHorizontally);
         void processSnapshot(YuvImage image, boolean consistentWithView, boolean flipHorizontally);
-        void dispatchOnVideoTaken(File file);
+        void dispatchOnVideoTaken(File file, boolean hasReachedMaxFileSize);
```
As above, we can't do this right now.

natario1 commented on cameraview/src/main/java/com/otaliastudios/cameraview/CameraView.java:
```diff
@@ -1335,6 +1335,24 @@ public boolean getPlaySounds() {
+     *
+     * @param maxFileSizeInBytes The maximum size of videos in bytes
+     */
+    public void setMaxFileSize(long maxFileSizeInBytes){
```
Can you rename to `setVideoMaxSize`?

chaitanyaraghav commented:
@natario1 Ya I agree. I was on the fence about that and did not quite like changing the signature. How about adding a new callback method to notify file size limits? That way only people requiring a notification can override it.

natario1 commented:
That would be better but I prefer not doing it. You know, there is a lot of video metadata which would be cool to pass around, but we can't have a callback for everything only to deprecate them later.

The good solution is really to change the signature as you did with a `VideoResult` object but I can only do that in a major release like `v2`. So, please be patient. You can easily call `file.length()` for now.

chaitanyaraghav commented:
@natario1 Ya makes sense to me. I reverted the changes to the method signature. Please take a look! 
I'll stick with `file.length()` for now :)

natario1 commented:
Looks good to me. I will add tests for this later. Two things,

- can you add a `cameraVideoMaxSize` XML attribute? It's just a few lines in the view constructor
- can you add some info in `README.md`? You can use the `Camera Controls` table where all other attrs are.

If you don't want to, no problem.

natario1 commented:
And, since all video API use `capturing`, can you rename isRecordingVideo to `isCapturingVideo`? Thanks!

chaitanyaraghav commented:
@natario1 Sure, I'll make the changes

Johnson145 commented:
I'm afraid this feature is a bit problematic, isn't it? I was thinking about implementing the same for the `setMaxDuration` part to get a constant video length as requested in #65. However, [according to the docs](https://developer.android.com/reference/android/media/MediaRecorder.html) both events will result in: `Stopping happens asynchronously, there is no guarantee that the recorder will have stopped by the time the listener is notified.` So we may inform the user about a captured video which isn't available yet, right? Furthermore,  your current implementation stops the video a second time, although the media recorder itself stopped it already. This could cause troubles, too, couldn't it?

Johnson145 commented:
A bit hacky, but maybe you could listen to the `MEDIA_RECORDER_INFO_MAX_FILESIZE_APPROACHING` event instead? Then we were able to keep control about stopping the video. You may internally scale the max file size that is given by the user such that the `MEDIA_RECORDER_INFO_MAX_FILESIZE_APPROACHING` event is triggered after 90% of the scaled size has reached, but 100% of the unscaled one.

natario1 commented:
@Johnson145 I have not tried of course but I think the two issues solve each other.

When we get the callback the recorder is *about to auto-stop*. We anticipate and stop it, then notify the user. If it stops itself before we do, there should be no problem because our stopping is in a `try-catch` block.

Does this make sense to you?

The best thing would be to try but I'm not able right now. Maybe after it's merged. If it proves to be flacky, we can do the trick with 90%, seems good!

Johnson145 commented:
May work, I don't know. Depends on the question whether the `mMediaRecorder.stop()` call is thread safe. If one of the two calls is guaranteed to finish its job, you're right.  I'm afraid it could be possible that our second call is failing, because the first one is still running (but not yet finished). In that case we would just continue after the `try-catch` and inform the user about the video to early. In each case, I guess problems are very unlikely, but who knows whether they are impossible..

Sure, we could try it first!

Johnson145 commented:
Maybe it's even a good idea to give it a try first. If it proves to work, we could implement the `MEDIA_RECORDER_INFO_MAX_DURATION_REACHED` part, too. Unfortunately, the latter does not have `MEDIA_RECORDER_INFO_MAX_DURATION_APPROACHING` version.

natario1 commented:
Yes, given how unfriendly MediaRecorder is, issues might be likely. The max duration feature would be cool as well if you want to work on it! We could then deprecate startCapturingVideo(File, long).

natario1 commented:
@chaitanyaraghav are you planning to work on this soon? There's no hurry of course, I want to know if I should release 1.4.1 today or wait for you. This Sunday I'll leave for a while so it will be hard for me to add tests etc. Thanks so much!

chaitanyaraghav commented:
@natario1 It may be better to push this to the next release. I'm currently out of town and I won't be able to make the changes until I get back this Sunday.

natario1 commented:
That's fine man 👍  I'll release today then!

chaitanyaraghav commented:
@natario1 Changed the method name, added xml attribute for video size and updated the documentation. 

The 90% trick seems good but I have not made that change yet. The current implementation appears to work fine in my testing. If it turns out to be flaky, I'll make the change to scale the max size. 

Thanks for the patience!

natario1 commented:
Thanks!
